source/cpuid/cpuid.go

```diff
@@ -16,8 +16,100 @@ limitations under the License.
 
 package cpuid
 
+import (
+	"sigs.k8s.io/node-feature-discovery/source"
+)
+
+// Configuration file options
+type NFDConfig struct {
+	AttributeBlacklist []string `json:"attributeBlacklist,omitempty"`
+	AttributeWhitelist []string `json:"attributeWhitelist,omitempty"`
+}
+
+var Config = NFDConfig{
+	AttributeBlacklist: []string{
+		"BMI1",
+		"BMI2",
+		"CLMUL",
+		"CMOV",
+		"CX16",
+		"ERMS",
+		"F16C",
+		"HTT",
+		"LZCNT",
+		"MMX",
+		"MMXEXT",
+		"NX",
+		"POPCNT",
+		"RDRAND",
+		"RDSEED",
+		"RDTSCP",
+		"SGX",
+		"SSE",
+		"SSE2",
+		"SSE3",
+		"SSE4.1",
+		"SSE4.2",
+		"SSSE3",
+	},
+	AttributeWhitelist: []string{},
+}
+
 // Source implements FeatureSource.
 type Source struct{}
 
+// Filter function for cpuid labels
+var filter *keyFilter
+
 // Name returns an identifier string for this feature source.
 func (s Source) Name() string { return "cpuid" }
+
+func (s Source) Discover() (source.Features, error) {
+	// Get the cpu features as strings
+	features := source.Features{}
+
+	if filter == nil {
+		initFilter()
+	}
+
+	for _, f := range getFeatures() {
+		if filter.unmask(f) {
+			features[f] = true
+		}
+	}
+	return features, nil
+}
+
+type keyFilter struct {
+	keys      map[string]bool
+	whitelist bool
+}
+
+func initFilter() {
+	newFilter := keyFilter{keys: map[string]bool{}}
+	if len(Config.AttributeWhitelist) > 0 {
+		for _, k := range Config.AttributeWhitelist {
+			newFilter.keys[k] = true
+		}
+		newFilter.whitelist = true
+	} else {
+		for _, k := range Config.AttributeBlacklist {
+			newFilter.keys[k] = true
+		}
+		newFilter.whitelist = false
+	}
+	filter = &newFilter
+}
+
+func (f keyFilter) unmask(k string) bool {
+	if f.whitelist {
```

Comment:
this can be simplified to one if. you need to return true when whitelist and ok is true or when whitelist and ok are false. Thus,
```go
if _, ok := f.keys[k]; f.whitelist == ok {
...
}
```

Reply:
True  (I  didn't even think about that), but, I think think that this optimization clearly worsens the readability of the code. In the future, it  may be  hard to understand the logic  behind `f.whitelist == ok`. So, I'd leave it as it is (and let the compiler do the optimization)

Reply:
@kad ping